core/src/main/scala/org/apache/spark/status/AppStatusStore.scala

```diff
@@ -222,29 +223,20 @@ private[spark] class AppStatusStore(
     val indices = quantiles.map { q => math.min((q * count).toLong, count - 1) }
 
     def scanTasks(index: String)(fn: TaskDataWrapper => Long): IndexedSeq[Double] = {
-      Utils.tryWithResource(
-        store.view(classOf[TaskDataWrapper])
-          .parent(stageKey)
-          .index(index)
-          .first(0L)
-          .closeableIterator()
-      ) { it =>
-        var last = Double.NaN
-        var currentIdx = -1L
-        indices.map { idx =>
-          if (idx == currentIdx) {
-            last
-          } else {
-            val diff = idx - currentIdx
-            currentIdx = idx
-            if (it.skip(diff - 1)) {
-              last = fn(it.next()).toDouble
-              last
-            } else {
-              Double.NaN
-            }
-          }
-        }.toIndexedSeq
+      val quantileTasks = store.view(classOf[TaskDataWrapper])
```

Comment:
This used to get an iterator. I wonder if it's safe to materialize the whole collection? it seems to take care to use the iterator to skip over things for example.

Reply:
Yes. That was the original intend. unfortunately after converting to the scala collection, the skip() functionality is not there. Also the kvstore doesn't have any filter API to filter the "success" tasks.

The PR was for reducing the computational time for loading  the stagePage from the diskStore ( for history server), by avoiding in memory sorting.

Reply:
Yes. If we do, "if (status == "SUCCESS")" for every iterator value, we can't do the skip function.
Becuase, earlier we know the exact index we need to take. ie. we can directly skip to 25th percentile, 50th percentile and so on.  Now, we don't know which index has the 25th percentile of the "SUCCESS" value, unless we iterate each.

Otherwise, we have to filter the "SUCCESS" the tasks  prior, like I have done in the PR.

Reply:
Hm, I think this reintroduces the problem that this code fixed though. What about adding some kind of filter() method to KVStoreView? it could configure an optional `Predicate<T>` that is applied to `iterator()`. Commons Collections has a `FilterIterator` for this.

Reply:
CC @vanzin for the proposed new method.

Reply:
There's a comment at the top of this method that explains why `skip` was used. It avoids deserialization of data that is not needed here, which can get quite expensive. IIRC there's about 26 metrics, and the updated code means deserializing `26 * numberOfTasks` task instances from the disk store. With large stages that can become really slow.

Try creating a large stage (e.g. `sc.parallelize(1 to 100000, 100000).count()`), loading the resulting event log through the history server, and checking how long it takes to load the stage page the first time. The goal is to make it reasonably fast (I think it's currently in the 4-5s range). Too slow and it makes the page not very usable.

If this makes that too slow, perhaps loading all the successful task into memory might be an ok workaround. It sucks (there is a spike in memory usage) but shouldn't be too bad (guessing 256 bytes per `TaskDataWrapper` object, 100k tasks ~ 24Mb which doesn't sound horrible).

Reply:
Thank you @vanzin for the review. I will check the time with large number of tasks.

Reply:
@vanzin Yes. It seems, loading the stage page take lot more time if we enable disk store. 
InMemory store seems okay.

Reply:
Pardon if I'm missing something, but does this avoid the problem? everything is loaded, and then it's filtered.

Reply:
@srowen  Yes. everything is loaded in "sorted" order based on index, and then we do filtering. For In memory case, this doesn't cause any issue. but for diskStore extra de serialization overhead is there. 

May be one possible solution can be, for diskStore case, bring only first time and sort  based on the corresponding indices to compute the quantiles.

If the solution seems complicated, then we can tell the user that, summary metrics display the quantile summary of all the tasks, instead of completed.

correct me if I am wrong

Reply:
I see so there's no real way to do it right without deserializing everything because that's the only way to know for sure what's successful.

Leaving it as-is isn't so bad; if failed tasks are infrequent then the quantiles are still about right.

I can think of fudges like searching ahead from the current index for the next successful task and using that. For the common case that might make the quantiles a little better. Iterate ahead, incrementing idx, and give up when hitting the next index.

Hm @vanzin is that too complex vs taking the hit and deserializing it all? or just punt on this?

Reply:
I'm not sure the fudging would give you much better results. Perhaps a little bit...

As for the correct solution, I don't see a lot of good options.

- loading all successful tasks in memory will be a little slow and memory-intensive for large stages

Using my 100k stage example, deserializing everything, even once, will stlil be a little expensive (deserializing 100k tasks vs. deserializing `metricCount * quantileCount` tasks which is 100 tasks and change in the UI case).

- changing the way the indexing works, so that you can index by specific metrics for successful and failed tasks differently, would be tricky, and also would require changing the disk store version (to invalidate old stores).

- maybe something else I'm not able to come up with at the moment.

Unless someone has a better idea, maybe just using this code when using an in-memory store is an ok temporary "fix".